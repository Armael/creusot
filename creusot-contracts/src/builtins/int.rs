use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int;

impl PartialEq for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "eq_int"]
    fn eq(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ne_int"]
    fn ne(&self, _: &Int) -> bool {
        panic!()
    }
}

impl PartialOrd for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "le_int"]
    fn le(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "ge_int"]
    fn ge(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "lt_int"]
    fn lt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "gt_int"]
    fn gt(&self, _: &Int) -> bool {
        panic!()
    }

    #[creusot::spec::no_translate]
    fn partial_cmp(&self, _: &Self) -> Option<std::cmp::Ordering> {
        panic!()
    }
}

impl From<i32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "i32_to_int"]
    fn from(_: i32) -> Self {
        panic!()
    }
}

impl From<u32> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "u32_to_int"]
    fn from(_: u32) -> Self {
        panic!()
    }
}

impl From<usize> for Int {
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "usize_to_int"]
    fn from(_: usize) -> Self {
        panic!()
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::spec::no_translate]
    #[rustc_diagnostic_item = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
