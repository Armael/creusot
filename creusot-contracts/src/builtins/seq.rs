use creusot_contracts_proc::*;

use crate::Int;

#[creusot::builtins = "seq.Seq.seq"]
pub struct Seq<T>(std::marker::PhantomData<T>);

impl<T> Seq<T> {
    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.get"]
    pub fn index(self, _: Int) -> T {
        std::process::abort()
    }

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.length"]
    pub fn len(self) -> Int {
        std::process::abort()
    }

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.set"]
    pub fn set(self, _: Int, _: T) -> Self {
        std::process::abort()
    }

    #[trusted]
    #[logic]
    #[creusot::builtins = "seq.Seq.snoc"]
    pub fn push(self, _: T) -> Self {
        std::process::abort()
    }
    #[predicate]
    pub fn permutation_of(self, o: Self) -> bool {
        self.permut(o, 0, self.len())
    }

    #[trusted]
    #[predicate]
    #[creusot::builtins = "seq.Permut.permut"]
    pub fn permut(self, _: Self, _: Int, _: Int) -> bool {
        std::process::abort()
    }
}
