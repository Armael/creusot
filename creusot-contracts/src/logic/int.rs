use super::model::*;
use crate as creusot_contracts;
use crate::logic::*;
use creusot_contracts_proc::*;

use std::ops::*;

#[rustc_diagnostic_item = "creusot_int"]
pub struct Int(*mut ());

impl WellFounded for Int {}

impl Int {
    #[trusted]
    #[logic]
    #[creusot::builtins = "int.Power.power"]
    pub fn pow(self, _: Int) -> Int {
        absurd
    }
}

impl From<u8> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "prelude.UInt8.to_int"]
    fn from(_: u8) -> Self {
        absurd
    }
}
impl Model for u8 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "prelude.UInt8.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<u32> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn from(_: u32) -> Self {
        absurd
    }
}
impl Model for u32 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<u64> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn from(_: u64) -> Self {
        absurd
    }
}
impl Model for u64 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i64> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn from(_: i64) -> Self {
        absurd
    }
}
impl Model for i64 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<i32> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn from(_: i32) -> Self {
        absurd
    }
}
impl Model for i32 {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int32.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<usize> for Int {
    #[logic]
    #[trusted]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn from(_: usize) -> Self {
        absurd
    }
}
impl Model for usize {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.UInt64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl From<isize> for Int {
    #[logic]
    #[trusted]
    #[rustc_diagnostic_item = "isize_to_int"]
    fn from(_: isize) -> Self {
        absurd
    }
}
impl Model for isize {
    type ModelTy = Int;
    #[logic]
    #[creusot::builtins = "mach.int.Int64.to_int"]
    fn model(self) -> Self::ModelTy {
        Int::from(self)
    }
}

impl Add<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "add_int"]
    fn add(self, _: Int) -> Self {
        panic!()
    }
}

impl Sub<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "sub_int"]
    fn sub(self, _: Int) -> Self {
        panic!()
    }
}

impl Mul<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "mul_int"]
    fn mul(self, _: Int) -> Self {
        panic!()
    }
}

impl Div<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "div_int"]
    fn div(self, _: Int) -> Self {
        panic!()
    }
}

impl Rem<Int> for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "rem_int"]
    fn rem(self, _: Int) -> Self {
        panic!()
    }
}

impl Neg for Int {
    type Output = Int;
    #[creusot::no_translate]
    #[creusot::builtins = "neg_int"]
    fn neg(self) -> Self {
        panic!()
    }
}
