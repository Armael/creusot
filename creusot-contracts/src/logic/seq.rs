use crate::{
    invariant::*,
    logic::{ops::IndexLogic, Mapping},
    *,
};

#[trusted]
#[cfg_attr(creusot, creusot::builtins = "seq.Seq.seq")]
pub struct Seq<T: ?Sized>(std::marker::PhantomData<T>);

impl<T: ?Sized> Seq<T> {
    #[cfg(creusot)]
    #[trusted]
    #[creusot::builtins = "seq.Seq.empty"]
    pub const EMPTY: Self = { Seq(std::marker::PhantomData) };

    #[trusted]
    #[logic]
    #[open(self)]
    #[creusot::builtins = "seq.Seq.create"]
    pub fn new(_: Int, _: Mapping<Int, T>) -> Self {
        absurd
    }

    #[logic]
    #[open]
    pub fn get(self, ix: Int) -> Option<T>
    where
        T: Sized, // TODO : don't require this (problem: return type needs to be sized)
    {
        if 0 <= ix && ix < self.len() {
            Some(self.index_logic(ix))
        } else {
            None
        }
    }

    #[trusted]
    #[logic]
    #[open]
    #[creusot::builtins = "seq.Seq.get"]
    pub fn index_logic_unsized(self, _: Int) -> Box<T> {
        absurd
    }

    #[trusted]
    #[logic]
    #[open(self)]
    #[creusot::builtins = "seq.Seq.([..])"]
    pub fn subsequence(self, _: Int, _: Int) -> Self {
        absurd
    }

    #[trusted]
    #[logic]
    #[open(self)]
    #[creusot::builtins = "seq.Seq.singleton"]
    pub fn singleton(_: T) -> Self {
        absurd
    }

    #[logic]
    #[open]
    pub fn tail(self) -> Self {
        self.subsequence(1, self.len())
    }

    #[trusted]
    #[logic]
    #[open(self)]
    #[rustc_diagnostic_item = "seq_len"]
    #[creusot::builtins = "seq.Seq.length"]
    pub fn len(self) -> Int {
        absurd
    }

    #[trusted]
    #[logic]
    #[open(self)]
    #[creusot::builtins = "seq.Seq.set"]
    pub fn set(self, _: Int, _: T) -> Self {
        absurd
    }

    #[trusted]
    #[predicate]
    #[open(self)]
    #[creusot::builtins = "seq.Seq.(==)"]
    pub fn ext_eq(self, _: Self) -> bool {
        absurd
    }

    #[trusted]
    #[logic]
    #[open(self)]
    #[creusot::builtins = "seq.Seq.snoc"]
    pub fn push(self, _: T) -> Self {
        absurd
    }

    // FIXME: order of arguments
    #[trusted]
    #[logic]
    #[open(self)]
    #[creusot::builtins = "seq.Seq.cons"]
    pub fn push_front(_: T, _: Self) -> Self {
        absurd
    }

    #[trusted]
    #[logic]
    #[open(self)]
    #[creusot::builtins = "seq.Seq.(++)"]
    pub fn concat(self, _: Self) -> Self {
        absurd
    }

    #[trusted]
    #[logic]
    #[open(self)]
    #[creusot::builtins = "seq.Reverse.reverse"]
    pub fn reverse(self) -> Self {
        absurd
    }

    #[predicate]
    #[open]
    pub fn permutation_of(self, o: Self) -> bool {
        self.permut(o, 0, self.len())
    }

    #[trusted]
    #[predicate]
    #[open(self)]
    #[creusot::builtins = "seq.Permut.permut"]
    pub fn permut(self, _: Self, _: Int, _: Int) -> bool {
        absurd
    }

    #[trusted]
    #[predicate]
    #[open(self)]
    #[creusot::builtins = "seq.Permut.exchange"]
    pub fn exchange(self, _: Self, _: Int, _: Int) -> bool {
        absurd
    }

    #[open]
    #[predicate]
    pub fn contains(self, e: T) -> bool
    where
        T: Sized, // TODO : don't require this (problem: uses index)
    {
        pearlite! { exists<i : Int> 0 <= i &&  i < self.len() && self[i] == e }
    }

    #[open]
    #[predicate]
    pub fn sorted_range(self, l: Int, u: Int) -> bool
    where
        T: OrdLogic + Sized, // TODO : don't require this (problem: uses index)
    {
        pearlite! {
            forall<i : Int, j : Int> l <= i && i <= j && j < u ==> self[i] <= self[j]
        }
    }

    #[open]
    #[predicate]
    pub fn sorted(self) -> bool
    where
        T: OrdLogic + Sized, // TODO : don't require this (problem: uses index)
    {
        self.sorted_range(0, self.len())
    }
}

/// Ghost definitions
impl<T> Seq<T> {
    /// Constructs a new, empty `Seq<T>`.
    ///
    /// This is allocated on the ghost heap, and as such is wrapped in [`GhostBox`].
    ///
    /// # Example
    ///
    /// ```rust,creusot
    /// use creusot_contracts::{proof_assert, Seq};
    /// let ghost_seq = Seq::<i32>::new_ghost();
    /// proof_assert!(seq == Seq::new());
    /// ```
    #[trusted]
    #[pure]
    #[ensures(*result == Self::EMPTY)]
    pub fn new_ghost() -> GhostBox<Self> {
        #[cfg(creusot)]
        {
            loop {}
        }
        #[cfg(not(creusot))]
        {
            GhostBox(::std::marker::PhantomData)
        }
    }

    /// Returns the number of elements in the vector, also referred to as its 'length'.
    ///
    /// # Example
    /// ```rust,creusot
    /// use creusot_contracts::{ghost, proof_assert, Seq};
    ///
    /// let mut vec = Seq::new_ghost();
    /// let length = ghost! {
    ///     vec.push(1);
    ///     vec.push(2);
    ///     vec.push(3);
    ///     vec.len()
    /// };
    /// proof_assert!(length.inner() == 3);
    /// ```
    #[trusted]
    #[pure]
    #[ensures(result == self.len())]
    pub fn len_ghost(&self) -> Int {
        loop {}
    }

    /// Appends an element to the back of a collection.
    ///
    /// # Example
    /// ```rust,creusot
    /// use creusot_contracts::{ghost, proof_assert, Seq};
    ///
    /// let mut s = Seq::new_ghost();
    /// ghost! {
    ///     s.push(1);
    ///     s.push(2);
    ///     s.push(3);
    /// };
    /// proof_assert!(s[0] == 1i32 && s[1] == 2i32 && s[2] == 3i32);
    /// ```
    #[trusted]
    #[pure]
    #[ensures(^self == self.push(x))]
    pub fn push_ghost(&mut self, x: T) {
        let _ = x;
        loop {}
    }

    #[trusted]
    #[pure]
    #[ensures(^self == Seq::push_front(x, *self))]
    pub fn push_front_ghost(&mut self, x: T) {
        let _ = x;
        loop {}
    }

    /// Returns a reference to an element at `index` or `None` if `index` is out of bounds.
    ///
    /// # Example
    /// ```rust,creusot
    /// use creusot_contracts::{ghost, Int, proof_assert, Seq};
    ///
    /// let mut s = Seq::new_ghost();
    /// let gets = ghost! {
    ///     s.push(10);
    ///     s.push(40);
    ///     s.push(30);
    ///     let get1 = s.get(*Int::new(1));
    ///     let get2 = s.get(*Int::new(3));
    ///     (get1, get2)
    /// };
    /// proof_assert!(gets.inner().0 == Some(&40i32));
    /// proof_assert!(gets.inner().1 == None);
    /// ```
    #[trusted]
    #[pure]
    #[ensures(match self.get(index) {
        None => result == None,
        Some(v) => result == Some(&v),
    })]
    pub fn get_ghost(&self, index: Int) -> Option<&T> {
        let _ = index;
        loop {}
    }

    /// Returns a mutable reference to an element at `index` or `None` if `index` is out of bounds.
    ///
    /// # Example
    /// ```rust,creusot
    /// use creusot_contracts::{ghost, Int, proof_assert, Seq};
    ///
    /// let mut s = Seq::new_ghost();
    ///
    /// ghost! {
    ///     s.push(0);
    ///     s.push(1);
    ///     s.push(2);
    ///     if let Some(elem) = s.get_mut(*Int::new(1)) {
    ///         *elem = 42;
    ///     }
    /// };
    /// proof_assert!(s[0] == 0i32 && s[1] == 42i32 && s[2] == 2i32);
    /// ```
    #[trusted]
    #[pure]
    #[ensures(if self.get(index) == None {
        result == None && *self == ^self
    } else {
        match result {
            None => false,
            Some(r) => *r == (*self)[index] && ^r == (^self)[index]
        }
    })]
    #[ensures(forall<i: Int> i != index ==> (*self).get(index) == (^self).get(index))]
    #[ensures((*self).len() == (^self).len())]
    pub fn get_mut_ghost(&mut self, index: Int) -> Option<&mut T> {
        let _ = index;
        loop {}
    }

    /// Removes the last element from a vector and returns it, or `None` if it is empty.
    ///
    /// # Example
    /// ```rust,creusot
    /// use creusot_contracts::{ghost, proof_assert, Seq};
    ///
    /// let mut s = Seq::new_ghost();
    /// let popped = ghost! {
    ///     s.push(1);
    ///     s.push(2);
    ///     s.push(3);
    ///     s.pop()
    /// };
    /// proof_assert!(popped == Some(3i32));
    /// proof_assert!(s[0] == 1i32 && s[1] == 2i32);
    /// ```
    #[trusted]
    #[pure]
    #[ensures(if self.len() == 0 {
        *self == ^self && result == None
    } else {
        match result {
            None => false,
            Some(r) => *self == (^self).push(r)
        }
    })]
    pub fn pop_ghost(&mut self) -> Option<T> {
        loop {}
    }

    #[trusted]
    #[pure]
    #[ensures(if self.len() == 0 {
        *self == ^self && result == None
    } else {
        match result {
            None => false,
            Some(r) => *self == Seq::singleton(r).concat(^self)
        }
    })]
    pub fn pop_front_ghost(&mut self) -> Option<T> {
        loop {}
    }
}

impl<T: ?Sized> Seq<&T> {
    #[logic]
    #[open]
    #[trusted]
    #[creusot::builtins = "identity"]
    pub fn to_owned_seq(self) -> Seq<T> {
        pearlite! {absurd}
    }
}

impl<T> IndexLogic<Int> for Seq<T> {
    type Item = T;

    #[logic]
    #[trusted]
    #[open(self)]
    #[rustc_diagnostic_item = "seq_index"]
    #[creusot::builtins = "seq.Seq.get"]
    fn index_logic(self, _: Int) -> Self::Item {
        absurd
    }
}

impl<T: ?Sized> Invariant for Seq<T> {
    #[predicate(prophetic)]
    #[open]
    #[creusot::trusted_ignore_structural_inv]
    #[creusot::trusted_is_tyinv_trivial_if_param_trivial]
    fn invariant(self) -> bool {
        pearlite! { forall<i:Int> 0 <= i && i < self.len() ==> inv(self.index_logic_unsized(i)) }
    }
}
