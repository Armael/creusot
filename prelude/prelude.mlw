module CreusotPrelude
  type borrowed 'a = { current : 'a; final: 'a }

  let function ( *_ ) x = x.current
  let function ( ^_ ) x = x.final

  val borrow_mut (a : 'a) : borrowed 'a
    ensures { * result = a }
end
