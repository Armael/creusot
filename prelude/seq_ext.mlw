module SeqExt
use seq.Seq

function subsequence (s : seq 't) (i : int) (j : int) : seq 't = s[i..j]

function mk_empty () : seq 't = Seq.empty
end