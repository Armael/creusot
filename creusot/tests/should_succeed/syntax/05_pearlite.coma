
module C05Pearlite_HasLen3_Impl
  let%span s05_pearlite0 = "../05_pearlite.rs" 8 11 8 24
  
  let%span span1 = "" 0 0 0 0
  
  let%span span2 = "../../../../../creusot-contracts/src/std/slice.rs" 18 14 18 41
  
  let%span span3 = "../../../../../creusot-contracts/src/std/slice.rs" 19 14 19 42
  
  let%span span4 = "../../../../../creusot-contracts/src/model.rs" 90 8 90 31
  
  use prelude.prelude.UInt32
  
  use seq.Seq
  
  use prelude.prelude.Int
  
  use prelude.prelude.Slice
  
  use prelude.prelude.Borrow
  
  use seq.Seq
  
  use prelude.prelude.Slice
  
  use prelude.prelude.UIntSize
  
  use prelude.prelude.UIntSize
  
  constant max'0 : usize = [%#span1] (18446744073709551615 : usize)
  
  function shallow_model'1 (self : slice uint32) : Seq.seq uint32
  
  axiom shallow_model'1_spec : forall self : slice uint32 . ([%#span3] shallow_model'1 self = Slice.id self)
  && ([%#span2] Seq.length (shallow_model'1 self) <= UIntSize.to_int (max'0 : usize))
  
  function shallow_model'0 (self : slice uint32) : Seq.seq uint32 =
    [%#span4] shallow_model'1 self
  
  constant v  : slice uint32
  
  predicate has_len_3 [#"../05_pearlite.rs" 9 0 9 35] (v : slice uint32)
  
  goal vc_has_len_3 : true
end
module C05Pearlite_A_Type
  type t_a  =
    | C_A bool
  
  function any_l (_ : 'b) : 'a
  
  let rec t_a (input:t_a) (ret  (a:bool))= any
    [ good (a:bool)-> {C_A a = input} (! ret {a}) | bad (a:bool)-> {C_A a <> input} {false} any ]
    
  
  function a_a (self : t_a) : bool =
    match self with
      | C_A a -> a
      end
end
module C05Pearlite_StructInPearlite
  let%span s05_pearlite0 = "../05_pearlite.rs" 23 10 23 30
  
  use C05Pearlite_A_Type as A'0
  
  use prelude.prelude.Intrinsic
  
  let rec struct_in_pearlite (x:A'0.t_a) (return'  (ret:()))= (! bb0 [ bb0 = return' {_0} ] ) [ & _0 : () = any_l () ] 
    [ return' (result:())-> {[@expl:postcondition] [%#s05_pearlite0] x = A'0.C_A false} (! return' {result}) ]
    
end
module C05Pearlite_B_Type
  use prelude.prelude.UInt32
  
  use prelude.prelude.Int
  
  type t_b  =
    | C_B bool uint32
  
  function any_l (_ : 'b) : 'a
  
  let rec t_b (input:t_b) (ret  (field1:bool) (field2:uint32))= any
    [ good (field1:bool) (field2:uint32)-> {C_B field1 field2 = input} (! ret {field1} {field2})
    | bad (field1:bool) (field2:uint32)-> {C_B field1 field2 <> input} {false} any ]
    
end
module C05Pearlite_StructOrder
  let%span s05_pearlite0 = "../05_pearlite.rs" 31 10 31 48
  
  use C05Pearlite_B_Type as B'0
  
  use prelude.prelude.UInt32
  
  use prelude.prelude.Int
  
  use prelude.prelude.Intrinsic
  
  let rec struct_order (x:B'0.t_b) (return'  (ret:()))= (! bb0 [ bb0 = return' {_0} ] ) [ & _0 : () = any_l () ] 
    [ return' (result:())-> {[@expl:postcondition] [%#s05_pearlite0] x = B'0.C_B false (0 : uint32)}
      (! return' {result}) ]
    
end
module C05Pearlite_GhostClosure
  let%span s05_pearlite0 = "../05_pearlite.rs" 49 13 49 38
  
  use prelude.prelude.UInt32
  
  use map.Map
  
  use prelude.prelude.Snapshot
  
  use prelude.prelude.Intrinsic
  
  use prelude.prelude.Snapshot
  
  use prelude.prelude.Mapping
  
  use prelude.prelude.Int
  
  let rec ghost_closure (_1:()) (return'  (ret:()))= (! bb0
    [ bb0 = s0
      [ s0 =  [ &_x <- [%#s05_pearlite0] Snapshot.new (Mapping.from_fn (fun (a : uint32) -> a)) ] s1 | s1 = bb1 ]
      
    | bb1 = return' {_0} ]
    ) [ & _0 : () = any_l () | & _x : Snapshot.snap_ty (Map.map uint32 uint32) = any_l () ] 
    [ return' (result:())-> (! return' {result}) ]
    
end
module C05Pearlite_PearliteClosure
  use prelude.prelude.UInt32
  
  use prelude.prelude.Int
  
  use map.Map
  
  use prelude.prelude.Snapshot
  
  use prelude.prelude.Intrinsic
  
  let rec pearlite_closure (_x:Snapshot.snap_ty (Map.map uint32 bool)) (return'  (ret:()))= (! bb0
    [ bb0 = return' {_0} ]
    ) [ & _0 : () = any_l () ]  [ return' (result:())-> (! return' {result}) ] 
end
module C05Pearlite_Caller
  let%span s05_pearlite0 = "../05_pearlite.rs" 55 21 55 44
  
  use prelude.prelude.UInt32
  
  use map.Map
  
  use prelude.prelude.Snapshot
  
  use prelude.prelude.Intrinsic
  
  use prelude.prelude.Int
  
  let rec pearlite_closure'0 (_x:Snapshot.snap_ty (Map.map uint32 bool)) (return'  (ret:()))= any
    [ return' (result:())-> (! return' {result}) ]
    
  
  use prelude.prelude.Snapshot
  
  use prelude.prelude.Mapping
  
  let rec caller (_1:()) (return'  (ret:()))= (! bb0
    [ bb0 = s0
      [ s0 =  [ &_2 <- [%#s05_pearlite0] Snapshot.new (Mapping.from_fn (fun (_a : uint32) -> true)) ] s1 | s1 = bb1 ]
      
    | bb1 = s0 [ s0 = pearlite_closure'0 {_2} (fun (_ret':()) ->  [ &_1 <- _ret' ] s1) | s1 = bb2 ] 
    | bb2 = return' {_0} ]
    ) [ & _0 : () = any_l () | & _1 : () = any_l () | & _2 : Snapshot.snap_ty (Map.map uint32 bool) = any_l () ] 
    [ return' (result:())-> (! return' {result}) ]
    
end
module C05Pearlite_S_Type
  type t_s  =
    | C_S
  
  function any_l (_ : 'b) : 'a
  
  let rec t_s (input:t_s) (ret  )= any [ good -> {C_S  = input} (! ret) | bad -> {C_S  <> input} {false} any ] 
end
