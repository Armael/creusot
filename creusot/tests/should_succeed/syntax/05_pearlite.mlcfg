
module C05Pearlite_A_Type
  type c05pearlite_a_type  =
    | C05Pearlite_A_Type bool
    
  let function c05pearlite_a_type_A_a (self : c05pearlite_a_type) : bool = [@vc:do_not_keep_trace] [@vc:sp]
    match (self) with
      | C05Pearlite_A_Type a -> a
      end
end
module C05Pearlite_Solver_Interface
  use C05Pearlite_A_Type
  val solver [@cfg:stackify] (x : C05Pearlite_A_Type.c05pearlite_a_type) : ()
    ensures { [#"../05_pearlite.rs" 11 10 11 20] C05Pearlite_A_Type.c05pearlite_a_type_A_a x = C05Pearlite_A_Type.c05pearlite_a_type_A_a x }
    
end
module C05Pearlite_Solver
  use C05Pearlite_A_Type
  val solver [@cfg:stackify] (x : C05Pearlite_A_Type.c05pearlite_a_type) : ()
    ensures { [#"../05_pearlite.rs" 11 10 11 20] C05Pearlite_A_Type.c05pearlite_a_type_A_a x = C05Pearlite_A_Type.c05pearlite_a_type_A_a x }
    
end
module C05Pearlite_StructInPearlite_Interface
  use C05Pearlite_A_Type
  val struct_in_pearlite [@cfg:stackify] (x : C05Pearlite_A_Type.c05pearlite_a_type) : ()
    ensures { [#"../05_pearlite.rs" 14 10 14 30] x = C05Pearlite_A_Type.C05Pearlite_A_Type false }
    
end
module C05Pearlite_StructInPearlite
  use C05Pearlite_A_Type
  let rec cfg struct_in_pearlite [@cfg:stackify] [#"../05_pearlite.rs" 15 0 15 31] (x : C05Pearlite_A_Type.c05pearlite_a_type) : ()
    ensures { [#"../05_pearlite.rs" 14 10 14 30] x = C05Pearlite_A_Type.C05Pearlite_A_Type false }
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var x_1 : C05Pearlite_A_Type.c05pearlite_a_type;
  {
    x_1 <- x;
    goto BB0
  }
  BB0 {
    _0 <- ();
    return _0
  }
  
end
module C05Pearlite_B_Type
  use mach.int.Int
  use mach.int.UInt32
  type c05pearlite_b_type  =
    | C05Pearlite_B_Type bool uint32
    
end
module C05Pearlite_StructOrder_Interface
  use mach.int.Int
  use mach.int.UInt32
  use C05Pearlite_B_Type
  val struct_order [@cfg:stackify] (x : C05Pearlite_B_Type.c05pearlite_b_type) : ()
    ensures { [#"../05_pearlite.rs" 22 10 22 48] x = C05Pearlite_B_Type.C05Pearlite_B_Type false (0 : uint32) }
    
end
module C05Pearlite_StructOrder
  use mach.int.Int
  use mach.int.UInt32
  use C05Pearlite_B_Type
  let rec cfg struct_order [@cfg:stackify] [#"../05_pearlite.rs" 23 0 23 25] (x : C05Pearlite_B_Type.c05pearlite_b_type) : ()
    ensures { [#"../05_pearlite.rs" 22 10 22 48] x = C05Pearlite_B_Type.C05Pearlite_B_Type false (0 : uint32) }
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : ();
  var x_1 : C05Pearlite_B_Type.c05pearlite_b_type;
  {
    x_1 <- x;
    goto BB0
  }
  BB0 {
    _0 <- ();
    return _0
  }
  
end
module C05Pearlite_Field1IsTrue_Interface
  use C05Pearlite_B_Type
  predicate field1_is_true (x : C05Pearlite_B_Type.c05pearlite_b_type)
end
module C05Pearlite_Field1IsTrue
  use C05Pearlite_B_Type
  use mach.int.UInt32
  use mach.int.Int
  use mach.int.Int32
  predicate field1_is_true [#"../05_pearlite.rs" 27 0 27 31] (x : C05Pearlite_B_Type.c05pearlite_b_type) =
    [#"../05_pearlite.rs" 30 8 34 9] match (x) with
      | C05Pearlite_B_Type.C05Pearlite_B_Type (True) _ -> true
      | C05Pearlite_B_Type.C05Pearlite_B_Type _f field2 -> UInt32.to_int field2 = 0
      | _ -> false
      end
end
