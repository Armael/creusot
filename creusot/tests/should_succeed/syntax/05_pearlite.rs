extern crate creusot_contracts;

use creusot_contracts::*;

// Tests that we can use field access syntax in pearlite.

struct A {
    a: bool,
}

#[trusted]
#[ensures(x.a === x.a)]
pub fn solver(x: A) {}
