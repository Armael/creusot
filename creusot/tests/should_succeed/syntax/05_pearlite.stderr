warning: unused variable: `x`
  --> 05_pearlite.rs:39:9
   |
39 |     let x = ghost! { pearlite! { |a : u32| a  } };
   |         ^ help: if this is intentional, prefix it with an underscore: `_x`
   |
   = note: `#[warn(unused_variables)]` on by default

warning: unused variable: `x`
  --> 05_pearlite.rs:42:25
   |
42 | pub fn pearlite_closure(x: Ghost<Mapping<u32, bool>>) {}
   |                         ^ help: if this is intentional, prefix it with an underscore: `_x`

warning: unused variable: `a`
  --> 05_pearlite.rs:45:44
   |
45 |     pearlite_closure(ghost! { pearlite! { |a| true }});
   |                                            ^ help: if this is intentional, prefix it with an underscore: `_a`

warning: 3 warnings emitted

